### ISSUE TITLE
TabOrder plugin conflict prevents adding to queue (Knockout-Sortable library name collision)

### ISSUE THREAD

**Issue by L0rz** [smartin015/continuousprint#14]
i Can,t add annything to the Q after update from 1.3.1 to you version.
first i remove the old version and hold my setings,
after this teh Settings page has the right setting but the print q was empti, so i remove the plugin again whith remove of my settings restart and install it again.

But i still can´t add anything to the list.
![2022-03-14 14_26_05-0% - (operational) - Vivaldi](https://user-images.githubusercontent.com/7045891/158181238-2bc9ea18-e6b7-4c36-831c-eb5a8d41ca35.png)

**Comment by Tiri1i1** [comment]
Same issue observed on version 1.4.1, Print queue is always empty no matter my actions

**Comment by leveos** [comment]
i can comfirm the same problem. it appears to be a visual bug for me though. if i click the plus sign to add a file to the queue nothing showes up but if i then press start managing it starts to print the file that i added. still not usable in this state but i hope that helps narrow it down

**Comment by smartin015** [comment]
Hey all, thanks for the reports, and sorry this wasn't a flawless upgrade - I'm doing some development work to hopefully make updates a bit more smooth moving forward.

If any of you are using Chrome, can you:

1. Use `Ctrl+Shift+C` keyboard shortcut to open the dev tools
2. click on the "Console" tab,
3. right-click and "Save As" to get a log file
4. Upload it to a comment here?

(see also [here](https://stackoverflow.com/questions/7627113/save-the-console-log-in-chrome-to-a-file) for a visual)

Going by the comment from @leveos, there's probably something going on in the page itself that's causing failure to render the files.

**Comment by smartin015** [comment]
Actually - just pushed some additional logging to try and make the problem more obvious. Can you please follow the `.zip` style instructions [here](https://github.com/smartin015/continuousprint#add-the-plugin) to pull those changes, and try with that?

**Comment by leveos** [comment]
Hi just wanted to update you on the new issue. I now tried to reinstall 1.4.1 from the plugin manager just using the search function that’s built in. and it installs as unknown. Which shows up no where in my plugins.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

From: nathan ***@***.***>
Sent: Tuesday, March 15, 2022 2:49 PM
To: ***@***.***>
Subject: RE: [smartin015/continuousprint] Not Possible to add Items to Q (Issue #14)

Hi,

Sorry for the delay. I had to work my day job. I uninstalled version 1.4.1 and cleaned up data. Now reinstalled from the master.zip file from the URL you provided. But now there is no continuous print plugin at all.

Using another one of my printers that still has the 1.4.1 installed I pulled the data you asked for previously.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by Tiri1i1** [comment]
Same issue, I installed again the plugin using the .zip URL (https://github.com/smartin015/continuousprint/archive/master.zip) but it's tagged as Unknown and doesn't show up anymore. 
I did test beforehand, and the Queue system does not work. Only the first part added is printed, but when it's finished it does not switch to the next part.

**Comment by smartin015** [comment]
Just squashed the bug with installation - I verified in a test instance that I can install the plugin, reboot, and it shows up (after verifying that it showed up as unknown before). Try one more time?

**Comment by smartin015** [comment]
@leveos, I think your email reply ate your attachment. Can you post it to the github issue directly?

**Comment by leveos** [comment]
> @leveos, I think your email reply ate your attachment. Can you post it to the github issue directly?

here is the log file frome before. i will try the fresh install of the .zip now as well

[printer2.local-1647380846568.log](https://github.com/smartin015/continuousprint/files/8258018/printer2.local-1647380846568.log)
e

**Comment by leveos** [comment]
here is the console log fro the fresh install using the .zip
[printer4.local-1647390376697.log](https://github.com/smartin015/continuousprint/files/8258041/printer4.local-1647390376697.log)

**Comment by smartin015** [comment]
Thanks @leveos - that second zip is super helpful. Seems like octoprint is trying to load the tab before the plugin is initialized for some reason. I'm digging in now to see why that's happening.

**Comment by smartin015** [comment]
@leveos , can you confirm you're running OctoPrint v1.6.1 and a relatively recent version of your browser? (e.g. Chrome 99.0.4844.57)

**Comment by leveos** [comment]
octoprint 1.7.3
chrome 99.0.4844.74

**Comment by smartin015** [comment]
Thanks - I still haven't been able to reproduce the errors you're seeing on my environment. It also looks like you're using a lot of additional plugins. Can you list what you have installed? Perhaps there's a conflict somewhere...

**Comment by leveos** [comment]
yes extra stuff i use:
bed visualizer
consolidate temp control
continuous print
custom background
floating navbar
GCODE system commands
M73 progress
Navbar temprature plugin
PSU control
tab order
themeify

**Comment by leveos** [comment]
i could take a backup of my octopi instance and send it to you if that will help. then you could use it to restore one of yours

**Comment by leveos** [comment]
just a thought. i use PSU control, my pi stays on always and i power on my printer with a relay after that it connects to the printer. could the bug be caused by continuous print being loaded while the printer isnt connected?

**Comment by smartin015** [comment]
Hah! It's the tab order plugin.

Normally (i.e. in a more "vanilla" flavored octoprint) the continuous print tab is hidden in a drop-down, and not displayed on page load. With TabOrder installed, it's loaded first-thing. I can now reproduce the problem with TabOrder installed. Whew.

If it's not too much to ask, can you uninstall TabOrder on your instance and see if it loads? I'll work on a fix from my end, but I'd like to see if that at least mitigates the problem on your end.

**Comment by leveos** [comment]
will do that right now. i also have continuous print as my very first tab on page load as its the first thing i need to get started for work

**Comment by smartin015** [comment]
> will do that right now. i also have continuous print as my very first tab on page load as its the first thing i need to get started for work

That's really convenient. I think I'll adopt that workflow for my own setup... thanks for the inadvertent tip :)

> ok with it uninstalled it still happens to be my first tab for some reason. maybe lingering settings. but it is allowing me to add files to the queue now. would still like to have tab order installed if possible in the future as i can use it to hide unwanted tabs. but this is good

Wacky. Agreed this isn't a great solution, but at least we know how to mitigate the problem and I can start fixing it for real now that I can reproduce it. I'll take a look tomorrow as it's getting a bit late tonight and I haven't been able to land on a "quick fix" in just a few minutes. Thanks again for all your help!

**Comment by smartin015** [comment]
After some more debugging, it turns out there's two different projects called `knockout-sortable`, and that's causing the plugins to collide:

* Continuous Print uses https://github.com/SortableJS/knockout-sortablejs, which binds https://github.com/SortableJS/Sortable
* TabOrder uses https://github.com/rniemeyer/knockout-sortable, which binds https://jqueryui.com/sortable/

The TabOrder sortable implementation unfortunately [doesn't support nesting](https://github.com/rniemeyer/knockout-sortable/issues/204), which is used now extensively in continuousprint to allow reordering of jobs and of their contained items. Otherwise, it'd be possible to just migrate to their dependency as a quick (but brittle) workaround. 

I'm attempting to mutate the naming of the dependencies used in this plugin so that the naming differs, which should hopefully fix this.

**Comment by smartin015** [comment]
Just merged in the rename, and I've experimentally verified it works with TabOrder also installed. @leveos - when you have a moment, please give it a try again (uninstall & reinstall `1.4.1`) and let me know if it works for you.

**Comment by leveos** [comment]
> Just merged in the rename, and I've experimentally verified it works with TabOrder also installed. @leveos - when you have a moment, please give it a try again (uninstall & reinstall `1.4.1`) and let me know if it works for you.

@smartin015 I can confirm that it is working perfectly with tab order installed now. just set up a large queue i will update this if for some reason it doesn't move onto the second file but i see no issues as of now.

**Comment by leveos** [comment]
i just want to update you that continuous print is working flawlessly with my setup now.

I also want to add that i really appreciate the optomizations you have done. the plugin is much more responsive now to adding and removing items from the queue and the gui is much more compact so i can see more items at a glance.

**Comment by Tiri1i1** [comment]
I am sadly still having issues with the queue going forward. 
I am using this with Klipper on my CR-30, with the following G-Code between prints (same as before)
G28 X0 Y0 Z0
G91 ; Set relative for lift
G0 Z10 F200
G90 ;back to absolute positioning
G28 X0 Y0 Z0; home

The queue is showing up well on the Plugin tab, and it does start with the first model, however after it's finished it cannot continue to the 2nd one.

**Comment by smartin015** [comment]
> i just want to update you that continuous print is working flawlessly with my setup now.
> 
> I also want to add that i really appreciate the optomizations you have done. the plugin is much more responsive now to adding and removing items from the queue and the gui is much more compact so i can see more items at a glance.

Thanks @leveos, that really means a lot and encourages me to do more :) I'm going to bump the version to `1.4.2` and close out this bug now that we have multiple fixes confirmed.

> I am sadly still having issues with the queue going forward. I am using this with Klipper on my CR-30, with the following G-Code between prints (same as before) G28 X0 Y0 Z0 G91 ; Set relative for lift G0 Z10 F200 G90 ;back to absolute positioning G28 X0 Y0 Z0; home
> 
> The queue is showing up well on the Plugin tab, and it does start with the first model, however after it's finished it cannot continue to the 2nd one.

Sorry your setup is still misbehaving @Tiri1i1 - can you please open a [new bug report](https://github.com/smartin015/continuousprint/issues/new?assignees=&labels=bug&template=bug_report.md&title=) and complete all the steps in the template? 

It's not busywork - it actually helps me root-cause faster with the system bundle and logs attached, and what you're reporting sounds like a different issue vs this one. Thanks!

**Comment by leveos** [comment]
Hi,

Sorry for the delay. I had to work my day job. I uninstalled version 1.4.1 and cleaned up data. Now reinstalled from the master.zip file from the URL you provided. But now there is no continuous print plugin at all.

Using another one of my printers that still has the 1.4.1 installed I pulled the data you asked for previously.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows
